dotnet/parity/mock_upstream.py

```
@@ -0,0 +1,118 @@
+chain RPC/esplora pools. Every request is recorded to a JSONL log keyed by an
+X-Parity-Marker propagated... (markers actually ride inside bodies/queries since
+the proxies don't forward custom headers) — pairing is done by (path, body hash).
+Responses are deterministic echoes so both proxies see identical upstreams.
+"""
+import json
+import re
+import threading
+from http.server import BaseHTTPRequestHandler, ThreadingHTTPServer
+from urllib.parse import urlparse, parse_qsl
+
+LOG_PATH = "/tmp/claude-0/-root/f686fea4-ddd5-474b-b474-899ed779389c/scratchpad/parity/upstream-log.jsonl"
```

Ruff (0.15.20)

[error] Probable insecure usage of temporary file or directory: "/tmp/claude-0/-root/f686fea4-ddd5-474b-b474-899ed779389c/scratchpad/parity/upstream-log.jsonl"

(S108)